packages/zoe/src/contracts/atomicSwap.js

```diff
@@ -28,8 +28,11 @@ const start = zcf => {
     const { want, give } = firstSeat.getProposal();
 
     /** @type {OfferHandler} */
-    const matchingSeatOfferHandler = matchingSeat =>
-      swap(zcf, firstSeat, matchingSeat);
+    const matchingSeatOfferHandler = matchingSeat => {
+      const swapResult = swap(zcf, firstSeat, matchingSeat);
+      zcf.shutdown();
+      return swapResult;
```

Comment:
can you return things after calling shutdown?

Reply:
What kind of activities can happen after calling zcf.shutdown, since it is synchronously called?

Reply:
I'll admit this surprised me too, but I surmise that no more messages can be sent to objects in the vat. But this function is already executing, and has to return somehow. Returning a constant should be trivial. Returning a value that was previously computed and is in a variable should work. I dunno what else we could get away with.

@FUDCo ?  What should we expect here?

Reply:
Well, you can _send_ messages to objects in a terminated vat, but any messages that hit the vat after it has been terminated will bounce off with an error (i.e., the result will be rejected).  Terminating a vat will also have the effect of rejecting any previously existing promises the vat was the decider.  Any messages sent _by_ the vat prior to its termination will be delivered.

If any information needs to get out of zcf before shutdown, you need to arrange for that to happen.  For example, you could give zcf the ability to terminate itself (maybe it already does) and have take care of any unfinished business and them pull the trigger on itself by sending a message to vatAdmin.  However, what you are doing here seems completely legit.

Reply:
> am I correct in my understanding that the contract is running inside the zcf vat? 

Yep.

> which will have the effect of cutting the vat being terminated off from the world. Vat termination does not abort any execution in progress. Anything done in the same crank as your issuance of the terminate message to the vatAdmin will happen (such as sending messages, resolving promises, etc.). In addition, any messages addressed to the vat that were in the kernel run queue prior to termination will be processed since the termination message will be placed into the run queue behind those.

I think that's what I'm counting on. Calls out to Zoe will be delivered. Calls within ZCF are mostly  synchronous, so they'll have completed by the time the shutdown takes effect.

There are some calls back-and-forth between zoe and zcf to manage seats. @katelynsills we should think about whether those could leave any incomplete state that would be cut off by a termination. My instinct is that Everything that would affect payouts and results will already have taken place. But I suspect I'm not thinking about all possible invocations by a contract. Do we need to add some explicit guidance for contract writers about which invocations need to have been finalized before it's safe to call shutdown?

Reply:
I think that makes sense. Can we create a separate issue to think this through and analyze it?